test/unittests/utils/test_connection_live.py

```diff
@@ -1,4 +1,217 @@
-from dsp_tools.utils.connection_live import ConnectionLive
+# mypy: disable-error-code="method-assign"
+
+import json
+from dataclasses import dataclass
+from importlib.metadata import version
+from typing import Any, Callable, cast
+from unittest.mock import Mock, patch
+
+import pytest
+from requests import ReadTimeout, RequestException
+
+from dsp_tools.models.exceptions import PermanentConnectionError, UserError
+from dsp_tools.utils.connection_live import ConnectionLive, RequestParameters
+
+
+@dataclass
+class SessionMock:
+    responses: tuple[Any, ...]
+    counter = 0
+
+    def request(self, **kwargs: Any) -> Any:  # noqa: ARG002
+        response = self.responses[self.counter]
+        self.counter += 1
+        if isinstance(response, BaseException):
+            raise response
+        return response
+
+
+@dataclass
+class ResponseMock:
+    status_code: int
+    headers: dict[str, Any]
+    text: str
+
+
+def test_log_in_log_out() -> None:
+    con = ConnectionLive("http://example.com/")
+    con.post = Mock(return_value={"token": "token"})
+    con.login("root@example.com", "test")
+    assert con.post.call_args.kwargs["route"] == "/v2/authentication"
+    assert con.token == "token"
+    assert con.session.headers["Authorization"] == "Bearer token"
+    con.delete = Mock()
+    con.logout()
+    assert con.token is None
+    assert "Authorization" not in con.session.headers
+
+
+def test_log_in_bad_credentials() -> None:
+    con = ConnectionLive("http://example.com/")
+    con._log_response = Mock()
+    con.session.request = Mock(return_value=Mock(status_code=401))
+    with patch("dsp_tools.utils.connection_live.time.sleep") as sleep_mock:
+        with pytest.raises(UserError, match="Username and/or password are not valid"):
+            con.login("invalid@example.com", "wrong")
+        sleep_mock.assert_not_called()
```

Comment:
What is the function of this?

Reply:
I want to make sure that when you log in with bad credentials, there is no retry, and no sleeping. Instead, a UserError should be raised immediately.